util/logutil/general_log_logger.go

```diff
@@ -0,0 +1,108 @@
+// Copyright 2021 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package logutil
+
+import (
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/natefinch/lumberjack"
+	"go.uber.org/zap"
+	"go.uber.org/zap/zapcore"
+)
+
+const (
+	generalLogBatchSize = 1024
+)
+
+// GeneralLogEntry represents the fields in a general query log
+type GeneralLogEntry struct {
+	ConnID                 uint64
+	FnGetUser              func() string
+	FnGetSchemaMetaVersion func() int64
+	TxnStartTS             uint64
+	TxnForUpdateTS         uint64
+	IsReadConsistency      bool
+	CurrentDB              string
+	TxnMode                string
+	FnGetQuery             func(*strings.Builder) string
+}
+
+var fnGetQueryPool = sync.Pool{New: func() interface{} {
+	ret := strings.Builder{}
+	ret.Grow(128)
+	return &ret
+}}
+
+// GeneralLogManager receives general log entry from channel, and log or drop them as needed
+type GeneralLogManager struct {
+	logger          *zap.Logger
+	logEntryChan    chan *GeneralLogEntry
+	droppedLogCount uint64
+}
+
+func newGeneralLog(logger *zap.Logger) *GeneralLogManager {
+	gl := &GeneralLogManager{
+		logger:          logger,
+		logEntryChan:    make(chan *GeneralLogEntry, generalLogBatchSize),
+		droppedLogCount: 0,
+	}
+	go gl.startFormatWorker()
+	return gl
+}
+
+// startFormatWorker starts a log flushing worker that flushes log periodically or when batch is full
+func (gl *GeneralLogManager) startFormatWorker() {
+	for {
+		e := <-gl.logEntryChan
```

Comment:
We need to update the TiDB graceful shutdown, wait until everything in **this** channel is consumed, to avoid losing logs.

Reply:
Fixed, PTAL.